pkg/compactor/shuffle_sharding_grouper.go

```diff
@@ -102,21 +120,42 @@ func (g *ShuffleShardingGrouper) Groups(blocks map[ulid.ULID]*metadata.Meta) (re
 	// which we can parallelly compact.
 	var outGroups []*compact.Group
 
-	i := 0
+	// Check if this compactor is on the subring.
+	// If the compactor is not on the subring when using the userID as a identifier
+	// no plans generated below will be owned by the compactor so we can just return an empty array
+	// as there will be no planned groups
+	onSubring, err := g.checkSubringForCompactor()
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to check sub-ring for compactor ownership")
+	}
+	if !onSubring {
+		level.Info(g.logger).Log("msg", "compactor is not on the current sub-ring skipping user", "user", g.userID)
+		return outGroups, nil
+	}
+	// Metrics for the remaining planned compactions
+	g.remainingPlannedCompactions.Set(0)
+
 	for _, mainBlocks := range mainGroups {
 		for _, group := range groupBlocksByCompactableRanges(mainBlocks, g.compactorCfg.BlockRanges.ToMilliseconds()) {
 			// Nothing to do if we don't have at least 2 blocks.
 			if len(group.blocks) < 2 {
 				continue
 			}
 
-			// TODO: Use the group's hash to determine whether a compactor should be responsible for compacting that group
-			groupHash := hashGroup(group.blocks[0].Thanos.Labels["__org_id__"], group.rangeStart, group.rangeEnd)
+			groupHash := hashGroup(g.userID, group.rangeStart, group.rangeEnd)
 
-			groupKey := fmt.Sprintf("%v%d", groupHash, i)
-			i++
+			if owned, err := g.ownGroup(groupHash); err != nil {
+				level.Warn(g.logger).Log("msg", "unable to check if user is owned by this shard", "group hash", groupHash, "err", err, "group", group.String())
+				continue
+			} else if !owned {
+				level.Info(g.logger).Log("msg", "skipping group because it is not owned by this shard", "group_hash", groupHash)
+				continue
+			}
+
+			g.remainingPlannedCompactions.Inc()
```

Comment:
There was an comment from bboreham in the staled duplicate PR:  https://github.com/cortexproject/cortex/pull/4432#discussion_r709389625

Below is copy of the comment: 

>If the metric is scraped in the middle of this function, it looks like it will have some value between 0 and the final value. Is that useful?
Would it be better to count up then `Set()` once?

Reply:
I've addressed it.